### PR TITLE
[core] Optimize MAC address formatting to eliminate sprintf dependency

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

This PR optimizes MAC address formatting functions to use direct arithmetic conversion instead of `sprintf`/`snprintf`, resulting in:
- **~20x performance improvement** (117ns → 5.3ns for pretty format, 101ns → 5.4ns for compact format)
- **RAM savings on ESP8266** - Format strings like `"%02X:%02X:%02X:%02X:%02X:%02X"` and `"%02x%02x%02x%02x%02x%02x"` were being stored in RAM, consuming precious memory on this memory-constrained platform
- **Flash savings** on both ESP8266 (-8 bytes) and ESP32 (-88 bytes)
- Elimination of heavy sprintf formatting code dependency for MAC operations
- Cleaner code with common helper functions and no duplication

The optimization replaces sprintf-based MAC formatting with efficient nibble-to-hex character conversion using simple arithmetic operations. On ESP8266, this is particularly important as the format strings were consuming RAM rather than being stored in Flash/PROGMEM, making this optimization even more valuable for memory-constrained devices.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):** Builds on optimization approach from #10710

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** N/A (internal implementation detail, no user-facing changes)

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

### 1. Added Common Helper Functions in `core/helpers.h`
- `format_hex_char_lower(uint8_t v)` - Convert nibble to lowercase hex char (a-f)
- `format_hex_pretty_char(uint8_t v)` - Convert nibble to uppercase hex char (A-F)
- `format_mac_addr_upper(const uint8_t *mac, char *output)` - Format as XX:XX:XX:XX:XX:XX
- `format_mac_addr_lower_no_sep(const uint8_t *mac, char *output)` - Format as xxxxxxxxxxxx

### 2. Updated Implementations
- `core/helpers.cpp`:
  - `format_mac_address_pretty()` - Now uses `format_mac_addr_upper()`
  - `get_mac_address()` - Now uses `format_mac_addr_lower_no_sep()`
  - `format_hex()` - Now uses `format_hex_char_lower()`
  - Kept `format_hex_pretty_char` as-is for backwards compatibility

- `ethernet/ethernet_component.cpp`:
  - `get_eth_mac_address_pretty()` - Now uses `format_mac_addr_upper()`

### ESP8266
Before:
```
RAM:   [====      ]  36.6% (used 29984 bytes from 81920 bytes)
Flash: [===       ]  34.1% (used 348783 bytes from 1023984 bytes)
```

After:
```
RAM:   [====      ]  36.6% (used 29968 bytes from 81920 bytes)  // -16 bytes
Flash: [===       ]  34.1% (used 348775 bytes from 1023984 bytes)  // -8 bytes
```

### ESP32
Before:
```
RAM:   [          ]   4.9% (used 16084 bytes from 327680 bytes)
Flash: [===       ]  27.4% (used 502006 bytes from 1835008 bytes)
```

After:
```
RAM:   [          ]   4.9% (used 16084 bytes from 327680 bytes)  // No change
Flash: [===       ]  27.4% (used 501918 bytes from 1835008 bytes)  // -88 bytes
```

## Performance Benchmark Results

Benchmark with 10,000 iterations on 1,000 test MAC addresses:

### Pretty Format (XX:XX:XX:XX:XX:XX)
```
sprintf                  : mean=117.14 ns (±5.90)
arithmetic               : mean=5.30   ns (±0.62) | 22.1x faster
lookup table             : mean=5.29   ns (±0.43) | 22.1x faster
```

### Compact Format (xxxxxxxxxxxx)
```
sprintf                  : mean=100.59 ns (±4.50)
arithmetic               : mean=5.41   ns (±0.20) | 18.6x faster
lookup table             : mean=5.13   ns (±0.33) | 19.6x faster
```

The arithmetic approach matches lookup table performance while avoiding memory overhead.

## Verification Tests

The attached `mac_formatting_tests.zip` contains:

1. **test_mac_format.cpp** - Basic correctness test verifying:
   - Edge cases (all zeros, all 0xFF)
   - Real MAC addresses
   - All 256 possible byte values

2. **test_mac_comprehensive.cpp** - Comprehensive test with:
   - Multiple real-world MAC addresses (Apple, VMware, etc.)
   - Boundary values and alternating patterns
   - 1000 random MAC addresses
   - Exhaustive validation

3. **benchmark_mac_format.cpp** - Performance benchmark comparing:
   - Original sprintf implementation
   - New arithmetic implementation
   - Alternative lookup table approaches
   - Detailed statistics with standard deviation and percentiles

All tests confirm 100% compatibility with identical output while delivering massive performance improvements.

## Example entry for `config.yaml`:

```yaml
# No configuration changes required - internal optimization only
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (see attached `mac_formatting_tests.zip`).

If user exposed functionality or configuration variables are added/changed:
  - N/A - Internal implementation change only, no user-facing changes.

## Additional Notes

This optimization applies the same proven approach from the MD5 hex conversion (PR #10710) to MAC address formatting, which is used much more frequently throughout the codebase for:
- Network status logging
- Device identification
- BLE device tracking
- Ethernet connectivity
- WiFi operations

The ~20x performance improvement will benefit all network-related operations while reducing memory usage.

[mac_formatting_tests.zip](https://github.com/user-attachments/files/22324171/mac_formatting_tests.zip)

### CODE REVIEW

**Comment by bdraco** [comment]
```python
#!/usr/bin/env python3
"""
Analyze ESP8266 firmware ELF file to find strings stored in RAM.

This script identifies strings that are in RAM sections (.data, .bss, .rodata)
rather than in flash sections (.irom0.text, .irom.text).
"""

import subprocess
import sys
import re
from pathlib import Path
from typing import List, Tuple, Dict, Set
import argparse


def run_command(cmd: List[str]) -> str:
    """Run a command and return its output."""
    try:
        result = subprocess.run(cmd, capture_output=True, text=True, check=True)
        return result.stdout
    except subprocess.CalledProcessError as e:
        print(f"Error running command: {' '.join(cmd)}", file=sys.stderr)
        print(f"Error: {e.stderr}", file=sys.stderr)
        sys.exit(1)
    except FileNotFoundError:
        print(f"Command not found: {cmd[0]}", file=sys.stderr)
        print("Please ensure you have the xtensa toolchain installed", file=sys.stderr)
        sys.exit(1)


def get_sections(elf_file: str) -> Dict[str, Tuple[int, int]]:
    """Get section addresses and sizes from ELF file."""
    sections = {}
    
    # Try to find the right objdump command
    objdump_cmd = None
    for cmd in ['xtensa-lx106-elf-objdump', 'xtensa-esp32-elf-objdump', 'objdump']:
        try:
            subprocess.run([cmd, '--version'], capture_output=True, check=True)
            objdump_cmd = cmd
            break
        except (subprocess.CalledProcessError, FileNotFoundError):
            continue
    
    if not objdump_cmd:
        print("Error: Could not find objdump command", file=sys.stderr)
        sys.exit(1)
    
    output = run_command([objdump_cmd, '-h', elf_file])
    
    # Parse section headers
    # Format: Idx Name          Size      VMA       LMA       File off  Algn
    section_pattern = r'^\s*\d+\s+(\S+)\s+([0-9a-fA-F]+)\s+([0-9a-fA-F]+)'
    
    for line in output.split('\n'):
        match = re.match(section_pattern, line)
        if match:
            name = match.group(1)
            size = int(match.group(2), 16)
            vma = int(match.group(3), 16)
            sections[name] = (vma, size)
    
    return sections


def get_strings_from_section(elf_file: str, section: str, min_length: int = 4) -> List[Tuple[int, str]]:
    """Extract strings from a specific section."""
    strings = []
    
    # Try to find the right objdump command
    objdump_cmd = None
    for cmd in ['xtensa-lx106-elf-objdump', 'xtensa-esp32-elf-objdump', 'objdump']:
        try:
            subprocess.run([cmd, '--version'], capture_output=True, check=True)
            objdump_cmd = cmd
            break
        except (subprocess.CalledProcessError, FileNotFoundError):
            continue
    
    if not objdump_cmd:
        return strings
    
    try:
        output = run_command([objdump_cmd, '-s', '-j', section, elf_file])
    except:
        return strings
    
    # Parse hex dump output
    current_string = bytearray()
    current_addr = 0
    string_start_addr = 0
    
    for line in output.split('\n'):
        # Lines look like: " 3ffef8a0 00000000 00000000 00000000 00000000  ................"
        match = re.match(r'^\s+([0-9a-fA-F]+)\s+((?:[0-9a-fA-F]{2,8}\s*)+)', line)
        if match:
            addr = int(match.group(1), 16)
            hex_data = match.group(2).strip()
            
            # Convert hex to bytes
            hex_bytes = hex_data.split()
            for hex_chunk in hex_bytes:
                # Handle both byte-by-byte and word formats
                if len(hex_chunk) == 2:
                    byte_val = int(hex_chunk, 16)
                    if 0x20 <= byte_val <= 0x7e:  # Printable ASCII
                        if not current_string:
                            string_start_addr = addr
                        current_string.append(byte_val)
                    else:
                        if byte_val == 0 and len(current_string) >= min_length:
                            # Found null terminator
                            strings.append((string_start_addr, current_string.decode('ascii', errors='ignore')))
                        current_string = bytearray()
                else:
                    # Handle multi-byte chunks (little-endian)
                    for i in range(0, len(hex_chunk), 2):
                        byte_val = int(hex_chunk[i:i+2], 16)
                        if 0x20 <= byte_val <= 0x7e:  # Printable ASCII
                            if not current_string:
                                string_start_addr = addr + i // 2
                            current_string.append(byte_val)
                        else:
                            if byte_val == 0 and len(current_string) >= min_length:
                                strings.append((string_start_addr, current_string.decode('ascii', errors='ignore')))
                            current_string = bytearray()
                
                addr += 1
    
    return strings


def analyze_symbols(elf_file: str) -> Dict[str, Tuple[str, int, int]]:
    """Get symbol information from ELF file."""
    symbols = {}
    
    # Try to find the right nm command
    nm_cmd = None
    for cmd in ['xtensa-lx106-elf-nm', 'xtensa-esp32-elf-nm', 'nm']:
        try:
            subprocess.run([cmd, '--version'], capture_output=True, check=True)
            nm_cmd = cmd
            break
        except (subprocess.CalledProcessError, FileNotFoundError):
            continue
    
    if not nm_cmd:
        return symbols
    
    output = run_command([nm_cmd, '-S', '--size-sort', elf_file])
    
    for line in output.split('\n'):
        parts = line.split()
        if len(parts) >= 4:
            addr = int(parts[0], 16)
            size = int(parts[1], 16) if parts[1] != '?' else 0
            sym_type = parts[2]
            name = ' '.join(parts[3:])
            
            # Filter for data symbols
            if sym_type in ['D', 'd', 'R', 'r', 'B', 'b']:
                symbols[name] = (sym_type, addr, size)
    
    return symbols


def main():
    parser = argparse.ArgumentParser(description='Analyze ESP8266 firmware for RAM-stored strings')
    parser.add_argument('elf_file', help='Path to firmware.elf file')
    parser.add_argument('--min-length', type=int, default=8, help='Minimum string length to report (default: 8)')
    parser.add_argument('--show-all', action='store_true', help='Show all sections, not just RAM')
    args = parser.parse_args()
    
    if not Path(args.elf_file).exists():
        print(f"Error: File not found: {args.elf_file}", file=sys.stderr)
        sys.exit(1)
    
    print(f"Analyzing: {args.elf_file}\n")
    
    # Get sections
    sections = get_sections(args.elf_file)
    
    # Define RAM sections for ESP8266
    ram_sections = {'.data', '.rodata', '.bss'}
    flash_sections = {'.irom0.text', '.irom.text', '.text'}
    
    print("=" * 80)
    print("SECTION ANALYSIS")
    print("=" * 80)
    print(f"{'Section':<20} {'Address':<12} {'Size':<12} {'Location'}")
    print("-" * 80)
    
    total_ram_usage = 0
    total_flash_usage = 0
    
    for name, (addr, size) in sorted(sections.items()):
        if name in ram_sections:
            location = "RAM"
            total_ram_usage += size
        elif name in flash_sections:
            location = "FLASH"
            total_flash_usage += size
        else:
            location = "OTHER"
        
        if args.show_all or name in ram_sections:
            print(f"{name:<20} 0x{addr:08x}   {size:>8} B   {location}")
    
    print("-" * 80)
    print(f"Total RAM sections size: {total_ram_usage:,} bytes")
    print(f"Total Flash sections size: {total_flash_usage:,} bytes")
    
    # Analyze strings in RAM sections
    print("\n" + "=" * 80)
    print("STRINGS IN RAM SECTIONS")
    print("=" * 80)
    
    all_ram_strings = []
    
    for section in ram_sections:
        if section in sections:
            strings = get_strings_from_section(args.elf_file, section, args.min_length)
            if strings:
                print(f"\nSection: {section}")
                print("-" * 40)
                for addr, string in sorted(strings):
                    if len(string) >= args.min_length:
                        # Clean up string for display
                        clean_string = string[:100] + ('...' if len(string) > 100 else '')
                        print(f"  0x{addr:08x}: \"{clean_string}\" (len={len(string)})")
                        all_ram_strings.append((section, addr, string))
    
    # Analyze symbols
    print("\n" + "=" * 80)
    print("LARGE DATA SYMBOLS IN RAM")
    print("=" * 80)
    
    symbols = analyze_symbols(args.elf_file)
    
    # Filter and sort by size
    ram_symbols = []
    for name, (sym_type, addr, size) in symbols.items():
        # Check if symbol is in a RAM range
        for section_name in ram_sections:
            if section_name in sections:
                section_start, section_size = sections[section_name]
                if section_start <= addr < section_start + section_size:
                    ram_symbols.append((name, sym_type, addr, size, section_name))
                    break
    
    # Sort by size (largest first)
    ram_symbols.sort(key=lambda x: x[3], reverse=True)
    
    print(f"\n{'Symbol':<40} {'Type':<6} {'Size':<10} {'Section'}")
    print("-" * 80)
    
    # Show top 20 largest symbols
    for name, sym_type, addr, size, section in ram_symbols[:20]:
        if size > 0:  # Only show symbols with known size
            print(f"{name[:39]:<40} {sym_type:<6} {size:>8} B  {section}")
    
    # Summary
    print("\n" + "=" * 80)
    print("SUMMARY")
    print("=" * 80)
    print(f"Total strings found in RAM: {len(all_ram_strings)}")
    total_string_bytes = sum(len(s) + 1 for _, _, s in all_ram_strings)  # +1 for null terminator
    print(f"Total bytes used by strings: {total_string_bytes:,}")
    
    # Identify potential optimization targets
    print("\n" + "=" * 80)
    print("POTENTIAL OPTIMIZATION TARGETS")
    print("=" * 80)
    
    # Find commonly repeated strings
    string_counts = {}
    for _, _, string in all_ram_strings:
        string_counts[string] = string_counts.get(string, 0) + 1
    
    repeated_strings = [(s, c) for s, c in string_counts.items() if c > 1]
    if repeated_strings:
        print("\nRepeated strings (could be deduplicated):")
        for string, count in sorted(repeated_strings, key=lambda x: x[1] * len(x[0]), reverse=True)[:10]:
            savings = (count - 1) * (len(string) + 1)
            clean_string = string[:50] + ('...' if len(string) > 50 else '')
            print(f"  \"{clean_string}\" - appears {count} times (potential savings: {savings} bytes)")
    
    # Find long strings that could be moved to PROGMEM
    long_strings = [(s, a, st) for s, a, st in all_ram_strings if len(st) >= 20]
    if long_strings:
        print(f"\nLong strings that could be moved to PROGMEM (>= 20 chars):")
        for section, addr, string in sorted(long_strings, key=lambda x: len(x[2]), reverse=True)[:10]:
            clean_string = string[:60] + ('...' if len(string) > 60 else '')
            print(f"  {section} @ 0x{addr:08x}: \"{clean_string}\" ({len(string)} bytes)")


if __name__ == '__main__':
    main()
```

tool for finding strings store in ram

` python3 /Users/bdraco/esphome/analyze_ram_strings.py /Users/bdraco/esphome/.esphome/build/log8266/.pioenvs/log8266/firmware.elf --min-length 10 `

File: esphome/core/helpers.cpp [hunk 1]
```diff
@@ -255,23 +255,22 @@ size_t parse_hex(const char *str, size_t length, uint8_t *data, size_t count) {
...
+  format_mac_addr_upper(mac, buf);
+  return std::string(buf);
 }
 
-static char format_hex_char(uint8_t v) { return v >= 10 ? 'a' + (v - 10) : '0' + v; }
 std::string format_hex(const uint8_t *data, size_t length) {
   std::string ret;
   ret.resize(length * 2);
   for (size_t i = 0; i < length; i++) {
-    ret[2 * i] = format_hex_char((data[i] & 0xF0) >> 4);
```
Comment (bdraco) [inline esphome/core/helpers.cpp:266]: Right shift automatically discards lower bits

File: esphome/core/helpers.cpp [hunk 2]
```diff
@@ -280,7 +279,7 @@ static std::string format_hex_pretty_uint8(const uint8_t *data, size_t length, c
   uint8_t multiple = separator ? 3 : 2;  // 3 if separator is not \0, 2 otherwise
   ret.resize(multiple * length - (separator ? 1 : 0));
   for (size_t i = 0; i < length; i++) {
-    ret[multiple * i] = format_hex_pretty_char((data[i] & 0xF0) >> 4);
```
Comment (bdraco) [inline esphome/core/helpers.cpp:283]: Right shift automatically discards lower bits

**Review by jesserockz — APPROVED** [review]
